### ISSUE TITLE
Memory usage during encode process: test data & graphs

### ISSUE THREAD

**Issue by ferdnyc** [OpenShot/openshot-qt#2811]
### Introduction
A few months ago, [I suggested some structured testing](https://github.com/OpenShot/openshot-qt/issues/2371#issuecomment-447192929) that might provide a more detailed picture of how OpenShot manages memory usage during the encoding process. Since nobody stepped up to provide any such data, I went through and at least made a start at collecting it myself. 

The seven test runs and resulting data are collected in this zip file:
[memory-test-dataset.zip](https://github.com/OpenShot/openshot-qt/files/3231643/memory-test-dataset.zip)

Hopefully this data will prove useful in examining and improving OpenShot's memory consumption, at least during the encode process.

### Test plan
Details on the test methodology and structure follow. (The contents of this section can also be found in `memory-test-notes.txt` in the zip file.)

I started with the standard HD test video, "Big Buck Bunny", pulled down
from YouTube using the youtube-dl utility. I selected video format 137
and audio format 256, aka...
```
137          mp4        1920x1080  1080p 4388k , avc1.640028, 30fps, video only, 154.89MiB
256          m4a        audio only DASH audio  196k , m4a_dash container, mp4a.40.5, 14.76MiB
```
Command to obtain the original video:
```sh
youtube-dl -f 137+256 aqz-KE-bpKQ
```

The resulting file:
```
78060661 Dec 12 05:08 BBB-full-1080p@30+192k.mp4
```
This I loaded into avidemux and split first into two ~5-minute clips:
```
77883511 May 29 00:31 BBB-5m-clip1_00:00:00-00:05:04.mp4
98428529 May 29 00:31 BBB-5m-clip2_00:05:04-00:10:16.mp4
```
Then five ~2-minute clips:
```
32280601 May 29 00:22 BBB-2m-clip1_00:00:00-00:02:02.mp4
29843096 May 29 00:23 BBB-2m-clip2_00:02:02-00:04:06.mp4
42106896 May 29 00:24 BBB-2m-clip3_00:04:06-00:06:12.mp4
37184325 May 29 00:25 BBB-2m-clip4_00:06:12-00:08:14.mp4
35536839 May 29 00:26 BBB-2m-clip5_00:08:14-00:10:16.mp4
```
Split points were at keyframes located at the indicated time signatures.

Project files (included) were set up as illustrated in the LayoutN.png
image files, containing some combinations of the Big Buck Bunny video
clip(s) laid out in various configurations. Each project file was set up
in OpenShot, saved, and the application quit. The .osp file was edited by
hand to remove the undo list from the history data, along with any other
adjustments that may have been needed.

Memory usage data was collected by running a fresh insance of openshot-qt
through the glibc memusage tool, logging to a binary data file:
```sh
memusage --data=ProjectN.dat openshot-qt ./ProjectN.osp
```
In this instance of OpenShot, the export process was immediately started
and allowed to run to completion, after which the application was again
exited (saving the updated project file when prompted).

The memusage data was then converted into a graph using memusagestat:
```sh
memusagestat -o ProjectN.png -t -T -x 1200 -y 800 ProjectN.dat
```

A couple of special notes on the individual test data files:

Project5.osp warrants some explanation. It consists of the full 10-minute
video clip, running concurrently with the 5-minute clips (sequentially).
All of the clips have Crop filter effects applied, so that one video
stream occupies the lower half of the screen, and the other occupies the
upper half of the screen. This way, both clips playing at any given time
are visible on-screen.

Project7.png is edited (from Project7orig.png) only to crop out some
extra area to the right of the graph, where the OpenShot process kept
running for several minutes after the encode job was completed. The
unnecessary "dead air" is trimmed off to make the graph more directly
comparable to the others in the set. (Note: It seems I also messed up
the background color, in the process.)

### Results

The following pairs of images (one for each of the 7 test projects) display a graph of process memory usage over time, alongside the project timeline layout which was encoded during the corresponding test.

##### Project1.osp
![Set1](https://user-images.githubusercontent.com/538020/58546198-dfd67580-81d2-11e9-8743-c4790284a228.png)
##### Project2.osp
![Set2](https://user-images.githubusercontent.com/538020/58546199-dfd67580-81d2-11e9-9f09-95115a68cf1c.png)
##### Project3.osp
![Set3](https://user-images.githubusercontent.com/538020/58546200-dfd67580-81d2-11e9-9b4c-9c45a4329b9e.png)
##### Project4.osp
![Set4](https://user-images.githubusercontent.com/538020/58546201-dfd67580-81d2-11e9-991f-f530fe2c6360.png)
##### Project5.osp
![Set5](https://user-images.githubusercontent.com/538020/58546202-dfd67580-81d2-11e9-8db7-cc6bf625367e.png)
##### Project6.osp
![Set6](https://user-images.githubusercontent.com/538020/58546203-e06f0c00-81d2-11e9-9973-881902f95203.png)
##### Project7.osp
![Set7](https://user-images.githubusercontent.com/538020/58546204-e06f0c00-81d2-11e9-84cf-e7db62c4c6e9.png)

### Analysis

Not stellar. There's definite room for improvement, as OpenShot is consuming a _lot_ of memory which is never released until the encode process completes. This means that as projects become more complex, the memory required to export the resulting video grows with _**every**_ element added to the timeline. Because very little of that memory is released _during_ the process, even non-concurrent clips add significant additional total memory requirements.

Project1 and Project3 show that if OpenShot nominally requires 233MB to export a 10-minute video from a single input clip, that memory requirement can nearly **double** simply by dividing the input video up into 5 separate clips, even though the input and output videos have the same total size and length.

Basically, looking at the `Project3.png` memusage graph, each rise in memory consumption represents the effect of introducing a new clip onto the timeline. The blue circles show how much memory is reclaimed from the previous clip, and the green circles show the **net** memory growth incurred by switching from clip A to clip B. (The two combined thus represent the total memory consumed by clip B during processing.)

![clip-growfree-detail](https://user-images.githubusercontent.com/538020/58547605-aeab7480-81d5-11e9-8a13-2da19791a544.png)

**Comment by ferdnyc** [comment]
Oh, details I should've included, for the record:

### Test system
All of this data was collected on my desktop system, specs as follows
```
	Manufacturer: Hewlett-Packard
	Product Name: HP Compaq 6200 Pro MT PC
Processor Information
	Family: Core i5
	Manufacturer: Intel
	Signature: Type 0, Family 6, Model 42, Stepping 7
	Version: Intel(R) Core(TM) i5-2400 CPU @ 3.10GHz
	Voltage: 1.2 V
	Max Speed: 3800 MHz
	Core Count: 4
	Core Enabled: 4
Memory Device
	Size: 2048 MB
	Locator: DIMM1
	Type: DDR3
	Speed: 1333 MT/s
	Manufacturer: Kingston        
	Part Number: KP223C-ELD        
Memory Device
	Size: 2048 MB
	Locator: DIMM2
	Type: DDR3
	Speed: 1333 MT/s
	Manufacturer: Nanya           
	Part Number: NT2GC64B8HC0NF-CG 
Memory Device
	Size: 2048 MB
	Locator: DIMM3
	Type: DDR3
	Speed: 1333 MT/s
	Manufacturer: Samsung         
	Part Number: M378B5673FH0-CH9  
Memory Device
	Size: 2048 MB
	Locator: DIMM4
	Type: DDR3
	Speed: 1333 MT/s
	Manufacturer: Samsung         
	Part Number: M378B5673FH0-CH9  
```
The system is running Fedora 30, with OpenShot using shared/system libraries wherever possible, installed from RPM Fusion packages:
```console
$ sudo dnf list installed \*openshot\*
Installed Packages
libopenshot.x86_64             0.2.3-2.20190406git101f25a.fc30 @rpmfusion-free  
libopenshot-audio.x86_64       0.1.8-2.20190405git7001b68.fc30 @rpmfusion-free  
libopenshot-audio-devel.x86_64 0.1.8-2.20190405git7001b68.fc30 @rpmfusion-free  
libopenshot-devel.x86_64       0.2.3-2.20190406git101f25a.fc30 @rpmfusion-free  
openshot.noarch                2.4.4-2.fc30                    @rpmfusion-free-updates
openshot-lang.noarch           2.4.4-2.fc30                    @rpmfusion-free-updates
python3-libopenshot.x86_64     0.2.3-2.20190406git101f25a.fc30 @rpmfusion-free  
```

### Export configuration
All exports were performed using the same profile, "HD 720p 30 fps (1280x720)" (meaning that OpenShot was downscaling the 1080p input video, intentionally as I wanted to give it something to do), and the videos were exported to the "MP4 (h.264)" target at "High" quality. (Resulting files were thus approx. 1.3 GB each.)

**Comment by SuslikV** [comment]
Let's me add my 2 cents. The export  to file controlled by FFmpeg. As soon as image was send to FFmpeg it is no longer needed and should be removed from the memory. Anyway, the export dialog window is modal and don't allow to you to use other functions of OpenShot, so all memory caching etc. is useless when export started. In general, only window in few MB is needed to feed the frames to FFmpeg as fast as possible, as far as I understand. All other stuff - is encoder's duty (it will fill encoder's buffer with required data, and will keep this data until it required by encoder's algorithm). The actual source image size and memory consumption value is important here (the output often set to 4:2:0, so by actual number of consumed RAM and output image resolution it is possible to analyze what data was lost in the memory - converted yuv or raw rgb).

Also, the memory consumption increases by using _Preview_ (it didn't fixed at value that it specified in the _Preferences_ - my own experience with OpenShot).

**Comment by ferdnyc** [comment]
> Let's me add my 2 cents. The export to file controlled by FFmpeg. As soon as image was send to FFmpeg it is no longer needed and should be removed from the memory. 

And in the very _immediate_ sense, that's clearly happening. The memory consumption is very flat throughout the lifetime of a given clip, indicating that all of the frame buffers that are being allocated are either being freed or re-used as the encode progresses from frame to frame to frame. On the _writer_ end, things seem to be well-managed.

But there's also clearly a _large_ allocation being made when each clip is loaded for processing, and almost **all** of that memory remains in use until the completion of the entire encode job (presumably getting freed only as a consequence of its entire thread terminating). Something that it would be extremely useful to track down.

> Anyway, the export dialog window is modal and don't allow to you to use other functions of OpenShot, so all memory caching etc. is useless when export started. In general, only window in few MB is needed to feed the frames to FFmpeg as fast as possible, as far as I understand. 

Well... yes and no. Because there could be multiple clips in play, at any given moment, at least one reader per (active) input video is required, plus the writer and its buffers, and if there are any other images (titles, animated title frames, etc) those as well. Not to mention, if any compositing or scaling is to be done, libopenshot coordinates that as well. And then (potentially) multiply (some of) that by the number of encode threads being run.

> All other stuff - is encoder's duty (it will fill encoder's buffer with required data, and will keep this data until it required by encoder's algorithm). The actual source image size and memory consumption value is important here (the output often set to 4:2:0, so by actual number of consumed RAM and output image resolution it is possible to analyze what data was lost in the memory - converted yuv or raw rgb).

When you're bloating by hundreds of megabytes, in massive leaps, I think that's perhaps a little too fine-grained. The leak doesn't appear to be happening at the buffer / individual-frame level (or it wouldn't happen all at once on file-open), it's a matter of something being loaded that never gets properly flushed/released. Quite possibly on the Python side.

> Also, the memory consumption increases by using _Preview_ (it didn't fixed at value that it specified in the _Preferences_ - my own experience with OpenShot).

Oh, no question there are other areas where memory use can be tightened as well. The reason I'm focusing on this one in particular is that it's the _**unavoidable**_ one.

1. UI bloat, such as growth due to an excessively-large undo/redo history, you can restart the application to clear up, at least temporarily. There's a workaround for that, even if it's not ideal.
2. Preview leaks, ditto. Plus, there's always the option to _not_ preview. (Obviously that's not a viable long-term approach, but my point is that even when Preview is causing memory leaks, it doesn't affect a session like the ones I ran for these tests: just load a project file and immediately hit Encode. You can still use OpenShot in that mode/capacity despite memory leaks due to the Preview functionality.)
3. But if the _**encode**_ process leaks, it's game over. There's absolutely NO workaround, nothing the user can do to reduce Encode's memory consumption _during_ the process of encoding a video. So however much memory OpenShot is going to consume when encoding a given project, there's no option other than to have that much memory available. And if it expands with the project input file count, that means that the amount of memory available to OpenShot is what controls the ability to encode projects of a given complexity _at all_.

**Comment by SuslikV** [comment]
All memory leaks that I seen (in my life ^_^) is comes form the c++ code.

Where in the code openshot doing compositing? Is it part of the libopenshot?

**Comment by ferdnyc** [comment]
> All memory leaks that I seen (in my life ^_^) is comes form the c++ code.

Heh, not wrong! :laughing: And I do think it's related to the C++ code... or, more specifically, the SWIG code. Because, all of the memory is _eventually_ freed — it just doesn't happen until the encode thread(s) is/are completed and the Python code finally destroys all of the handles and references it's holding on to.

I mean, one of the first things the export process does is create a _duplicate_ Timeline object. (A necessary evil, to deal with possible frame rate differences from the project timeline — things were _way_ worse before frame rate mismatches were handled.)
https://github.com/OpenShot/openshot-qt/blob/ba259c1ebc3ef346f1fab289a3702c6fb7fa0153/src/windows/export.py#L123-L137

And then eventually when the process starts, it loops over the requested frames one by one, `GetFrame()`s them from the Timeline reader (which applies the layering and effects to combine all  clips present at that frame), and passes them to the `FFMpegWriter`'s `WriteFrame()`.

https://github.com/OpenShot/openshot-qt/blob/ba259c1ebc3ef346f1fab289a3702c6fb7fa0153/src/windows/export.py#L848-L852

https://github.com/OpenShot/openshot-qt/blob/ba259c1ebc3ef346f1fab289a3702c6fb7fa0153/src/windows/export.py#L862-L869

Now, each Frame is just passed as a `std::shared_ptr< openshot:Frame >` which should be lightweight. But the `FFMpegReader` objects for all of the clips are going to be referenced in that copy of the `openshot.Timeline` until it's destroyed at the end of the export.

A lot of the libopenshot data structures rely on reference-counting to free up their memory allocations. One tripwire in working with PyQt5 and with SWIG is the danger of Python picking up references to C++ data structures that cause them to persist longer than normal.

> Where in the code openshot doing compositing? Is it part of the libopenshot?

It's implicit in that `self.timeline.GetFrame()` call, yeah. The Frame object that's returned (by reference) is the composited result of assembling the Timeline layer stack at that particular frame#.

**Comment by ferdnyc** [comment]
> Because, all of the memory is _eventually_ freed — it just doesn't happen until the encode thread(s) is/are completed and the Python code finally destroys all of the handles and references it's holding on to.

I kind of unintentionally glossed over this point. It's a critical aspect of the issue, though, because really the concern here _isn't_ a memory leak — at least, not by the strictest definition. The memory consumed during encoding does all (or close enough) **_eventually_** get freed. 

The issue is just that large chunks are being allocated, then held onto for far longer than should be necessary. It would greatly improve the encoder's memory efficiency if those blocks can be freed as soon as the encoder no longer needs them. Instead they're hanging around, accumulating with each additional input stream loaded, until the thread finalizes and everything finally gets deallocated en masse.

**Comment by ckirmse** [comment]
I think this will fix it: https://github.com/OpenShot/libopenshot/pull/240

**Comment by SuslikV** [comment]
Thank you for the explanation. Now I know a bit more about how OpenShot works.

**Comment by SuslikV** [comment]
@ferdnyc results images needs to be updated because of current changes in libopenshot.

**Comment by ferdnyc** [comment]
Indeed, I'll have to test how the fix @ckirmse provided in OpenShot/libopenshot#240 affects these results. I'm hopeful we'll see the same significant reduction in memory growth.

**Comment by ferdnyc** [comment]
So, to re-test I built a libopenshot with _just_ the Openshot/libopenshot#240 patch cherry-picked. (I didn't want to pick up the git HEAD and end up pulling in all of the hardware-encoding code that's been applied since v0.2.3.) Then I re-ran just Project1 and Project3 from my old dataset.

There's a _**marked**_ improvement. Project1 is virtually identical to the previous run, as one would expect (given the nature of the changes made in OpenShot/libopenshot#240):

#### Project1.osp
![Project1](https://user-images.githubusercontent.com/538020/58741731-7a5cd180-83ea-11e9-875c-c477525945ab.png)

But Project3 showed far better memory reclamation. There's still a _slight_ bit of growth from clip to clip, but now it's on the order of just a few megabytes total, rather than over 100 megabytes for each clip! And you can see where basically all of the memory from the previous clip is being freed, now. (That's visible in the _deep_ drops at the end of each clip, compared to the tiny little stubs I highlighted with blue circles in that previous graph detail image.)

#### Project3.osp
![Project3](https://user-images.githubusercontent.com/538020/58741733-821c7600-83ea-11e9-9a66-eaa297be2ece.png)

**Comment by ferdnyc** [comment]
(So, just to explicitly say it again: Thanks, @ckirmse — _huge_ win, with that fix!)

**Comment by SuslikV** [comment]
Big question is why the usage of RAM is so huge. In both examples only 1 frame is encoded each time (nothing to combine it with) and still it requires a lot of RAM for processing. The caching scheme for export not perfect in either way. Bug fix is OK but algorithm itself is bad. Maybe it works OK for preview but in export - it is not.

Also, I noticed that after the export task complete the RAM usage never returns back to initial value and from say 250MB it can temporary increase up to 1200MB during processing but will remain at 900MB when encoding complete. The setup includes effects, transitions, animations (speedup, scaling, position), clips sliced number of times to cut off unneeded parts and speed up some moments. Application was restarted before the export button was clicked (if it matter).

**Comment by ferdnyc** [comment]
> **Edit:** the screenshots is not perfect - I can imagine that it became even worse, because I see single flat line at maximum. 

They're not screenshots. They're automatically-scaled plots generated by the `memusagestat` tool, using collected data stored in binary data files by `memusage`. The graph does not extend beyond the visible scale.

> You need to place images side by side at same scale level.

That's not possible with `memusage`, the graphs are what they are. It only operates on the opaque binary data files it generates; one program run → one data file → one graph. You're right, they're not scaled the same, but they're also not scaled that differently. The max for each graph is shown in the red scale up the left vertical axis: It's 233M for Project1, 240M for Project3. That value corresponds to the max of the collected measurements.

The graphs were intended primarily to illustrate memory consumption over time, and to a lesser extent to compare the different runs in a broad-strokes sense. The scale's max-value label is sufficient for that purpose. 

From running both `openshot-qt` and `avidemux3_qt5`, I can say that the measured values _appear_ to be based on a single core, not too surprisingly. (`memusage` is useful because it requires no modifications to the program being profiled, but any such "easy" profiler won't have perfect accuracy.) Anyway, point is, on my system the application's _total_ memory consumption should be 4x the values shown by `memusage`, due to the 4-core CPU. Which is roughly what I observed: total memory consumption came in just under 1GB.

**Comment by ferdnyc** [comment]
I'm going to close this as addressed-for-now; I may re-run the tests against the latest code at some point in the future, perhaps even comparing the hardware- and software-encoding paths, but I can open a new issue at that time and just reference this one. Thanks again to @ckirmse for that quick fix on the severe memory bloat!

**Comment by ferdnyc** [comment]
I happened to have an opportunity to revisit this today, in testing OpenShot/libopenshot#375, so I re-ran just the `Project3.osp` test (third graph in the initial run, second graph in my later update) against the current development code.

The great news is, results show a not-insignificant improvement when compared to even the "post-OpenShot/libopenshot#240" runs, meaning memory consumption has continued to tighten up as we've been enhancing both OpenShot and libopenshot over the past few months.

Project3 graph with current `develop` branches of both libopenshot and OpenShot:

### Project3.osp
![Project3_baseline](https://user-images.githubusercontent.com/538020/69752736-dac19700-111f-11ea-82ad-476f81c244ac.png)

(Peak memory consumption, per thread, decreased from `240 MB` to `222 MB`, with **almost** no discernible leaks over the course of the encode process.)
